env/ipp_aes_ctr_provider.cc

```diff
@@ -0,0 +1,174 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#ifndef ROCKSDB_LITE
+
+#include "ipp_aes_ctr_provider.h"
+
+#include <regex>
+
+#endif
+
+namespace ROCKSDB_NAMESPACE {
+
+#ifndef ROCKSDB_LITE
+
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+
+IppAESCTRCipherStream::IppAESCTRCipherStream(IppsAESSpec* aes_ctx,
+                                             const char* init_vector)
+    : aes_ctx_(aes_ctx) {
+  init_vector_ = _mm_loadu_si128((__m128i*)init_vector);
+}
+
+Status IppAESCTRCipherStream::Encrypt(uint64_t fileOffset, char* data,
+                                      size_t dataSize) {
+  if (!dataSize) return Status::OK();
+
+  size_t index = fileOffset / kBlockSize;
+  size_t offset = fileOffset % kBlockSize;
+
+  Ipp8u ctr_block[kBlockSize];
+
+  __m128i counter = _mm_add_epi64(init_vector_, _mm_cvtsi64_si128(index));
+  Ipp8u* ptr_counter = (Ipp8u*)&counter;
+  for (size_t i = 0; i < kBlockSize; ++i)
+    ctr_block[i] = ptr_counter[kBlockSize - 1 - i];
+
+  IppStatus ipp_status = ippStsNoErr;
+  if (!offset) {
+    ipp_status = ippsAESEncryptCTR((Ipp8u*)(data), (Ipp8u*)data, dataSize,
+                                   aes_ctx_, ctr_block, kCounterLen);
+  } else {
+    Ipp8u zero_block[kBlockSize]{0};
+    ipp_status = ippsAESEncryptCTR(zero_block, zero_block, kBlockSize, aes_ctx_,
+                                   ctr_block, kCounterLen);
+    size_t n = std::min(kBlockSize - offset, dataSize);
+    for (size_t i = 0; i < n; ++i) data[i] ^= zero_block[offset + i];
+    memset(zero_block, 0, kBlockSize);
+
+    n = kBlockSize - offset;
+    if (dataSize > n) {
+      Ipp8u* ptr = (Ipp8u*)(data + n);
+      ipp_status = ippsAESEncryptCTR(ptr, ptr, dataSize - n, aes_ctx_,
+                                     ctr_block, kCounterLen);
+    }
+  }
+
+  if (ipp_status == ippStsNoErr) return Status::OK();
+
+  return Status::Aborted(ippcpGetStatusString(ipp_status));
+}
+
+Status IppAESCTRCipherStream::Decrypt(uint64_t fileOffset, char* data,
+                                      size_t dataSize) {
+  return Encrypt(fileOffset, data, dataSize);
+}
+
+#endif  // IPPCP
+
+Status IppAESCTRProvider::CreateProvider(const std::string& id,
+                                         IppAESCTRProvider** provider) {
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  // expected format: IPP_AES:<128|192|256>:key
+  if (!std::regex_match(id, std::regex("IPP_AES:(128|192|256):.+"))) {
+    return Status::NotSupported("Invalid provider ", id);
+  }
+
+  // key size starts at index 8, key starts at index 12
+  size_t key_size_bits = std::stoi(id.substr(8, 3), nullptr);
+  std::string key = id.substr(12);
+  if (key_size_bits != key.size() * 8) {
+    return Status::NotSupported("Key size mismatch in ", id);
+  }
+
+  IppAESCTRProvider* ptr_ipp = new IppAESCTRProvider();
+  ptr_ipp->ctx_size_ = 0;
```

Comment:
Why is the ctx_size_ not initialized in the constructor?

Reply:
Should be.